doc/arch/adr-002-geoprocessing-caching.md

```diff
@@ -0,0 +1,103 @@
+
+### How will the cache be invalidated?
+
+Since every cache entry is tagged with its type, if a certain raster is updated, we can remove all related cache entries. For example, if there are updates to `us-percent-slope-30m-epsg5070` raster, which is used in the [`nlcd_slope`](https://github.com/WikiWatershed/model-my-watershed/blob/a49cc115d18d2c67cbd0721cf62faa08e98d13fc/src/mmw/mmw/settings/base.py#L449) and [`slope`](https://github.com/WikiWatershed/model-my-watershed/blob/a49cc115d18d2c67cbd0721cf62faa08e98d13fc/src/mmw/mmw/settings/base.py#L462) requests, we could simply run
+
+```sql
+DELETE FROM geop_cache
+WHERE type IN ('nlcd_slope', 'slope');
+```
+
+If we'd like to delete all caches before a certain date, for example a deployment, we could do something like
+
+```sql
+DELETE FROM geop_cache
+WHERE created_at < '2017-04-25';
+```
+
+To refresh a specific shape we could do
+
+```sql
+DELETE FROM geop_cache
+WHERE name = 'boundary_huc08__1';
+```
+
+And finally, to delete the entire cache, we can just do
+
+```sql
+DELETE FROM geop_cache;
+```
+
+## Implementation Plan
+
+1. Create the `geop_cache` table with `name`, `type`, and `created_at` fields as described above.
+2. See if TR-55 geoprocessing, currently done via [`SummaryJob`](https://github.com/WikiWatershed/mmw-geoprocessing/blob/develop/summary/src/main/scala/SummaryJob.scala), can be done via [`MapshedJob`](https://github.com/WikiWatershed/mmw-geoprocessing/blob/develop/summary/src/main/scala/MapshedJob.scala), for consistency. And if so, rewrite the geoprocessing bits to use `MapshedJob` instead.
+3. Split [`nlcdSoilCensus`](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/mmw/settings/base.py#L388) into three requests: `nlcd`, `soil`, and `nlcd_soil`. The first two can be used for `/analyze` and will be much faster, while the second can be used for `/modeling/tr-55`.
+4. Update the geoprocessing submodule to take GeoJSON shape or WKAoI id, and geoprocessing type, and return the output. Update the Celery Tasks or Django Views which use the geoprocessing submodule to use the new interface.
+5. Add caching support to the geoprocessing submodule. All geoprocessing is done in two parts: `_start` and `_finish`. In the case of a cache hit, there should be a signal passed from `_start` to `_finish` that instructs it to fetch the value from the cache instead of `sjs_retrieve`.
+6. Update the UI to send WKAoI id instead of GeoJSON for predefined shapes. RWD and user defined shapes should still be sent as GeoJSON.
+
+## Consequences
+
+This will make the worst case scenario, a cache miss, slightly longer than it currently is, because we'll be checking the cache before doing the actual geoprocessing.
+
+We will have to monitor the characteristics of the cache table to see how it grows. If it gets too big too quickly, we may have to think of ways of routinely emptying the cache, or removing the least recently used entries, or something like that.
```

Comment:
I haven't worked with Redis enough to be entirely comfortable with it, but that shouldn't stop us from not choosing it if it is the better option. I am biased towards database simply because I know how to interact with it better (for cache invalidation operations, etc). If there are enough voices in support of Redis, I will switch to that.